### Loading dataframes

In [ ]:
import pandas as pd

### Selecting columns from dataframe

In [ ]:
df = pd.DataFrame({'a':[1,1,1,2,2,3,4,5],
                  'b':[10,10,11,20,20,30,40,50]})
df

In [ ]:
df['a']

In [ ]:
type(df), type(df['a'])

In [ ]:
df[['a','b']]

In [ ]:
type(df[['a']]), type(df['a'])

In [ ]:
df.columns

In [ ]:
df.index

In [ ]:
for i in df.index:
    print(i)

In [ ]:
df.columns[1]

### Check for duplicates

In [ ]:
df = pd.DataFrame({'a':[1,1,1,2,2,3,4,5],
                  'b':[10,10,11,20,20,30,40,50]})

In [ ]:
df

In [ ]:
df.duplicated()

In [ ]:
df.duplicated(subset=['a'])

In [ ]:
df.duplicated().sum()

### Drop duplicated rows

In [ ]:
df_no_duplicates = df.drop_duplicates()

In [ ]:
len(df), len(df_no_duplicates)

In [ ]:
df.drop_duplicates(inplace=True)

In [ ]:
len(df)

### Check for missing values

In [ ]:
df = pd.DataFrame({'a':[1,2,3,4],
                  'b':[10,None,30,40],
                  'c':[None,None,None,400]})
df

In [ ]:
df.isnull()

In [ ]:
df.isnull().any(axis=0)

In [ ]:
df.isnull().any(axis=1)

In [ ]:
df.isnull().sum()

In [ ]:
df.isnull().sum(axis=1)

### Drop rows with missing values

In [ ]:
df

In [ ]:
df_only_full_rows = df.dropna()
df_only_full_rows

In [ ]:
df_rows_where_b_is_not_missing = df.dropna(subset=['b'])
df_rows_where_b_is_not_missing

### Replace missing values

In [ ]:
df

In [ ]:
mean_b = df['b'].mean()
mean_b

In [ ]:
mean_b = df['b'].mean()
df_missing_b_replaced_with_mean = df.fillna(value={'b':mean_b})
df_missing_b_replaced_with_mean



In [ ]:
df_missing_b_replaced_with_mean_missing_c_replaced_with_zero = df.fillna(value={'b':mean_b,
                                                                                'c':0})
df_missing_b_replaced_with_mean_missing_c_replaced_with_zero

In [ ]:
df.fillna('Unknown', inplace=True)
df

### Describing numeric data

In [ ]:
df = pd.DataFrame({'a':[1,1,1,2,2,3,4,5],
                  'b':[10,10,11,20,None,None,40,50],
                  'c':['apple','apple','plum','pear','plum','apple','apple','apple']})

In [ ]:
df

In [ ]:
df.dtypes

In [ ]:
df['c'].dtype

In [ ]:
df.describe()

In [ ]:
df.max()

In [ ]:
df['a'].max()

### Describing non numeric data

In [ ]:
df

In [ ]:
df['c'].unique()

In [ ]:
df['c'].nunique()

In [ ]:
df['c'].value_counts()

## Exercise

### 1 - exercise
Load the datasets that you downloaded from the google drive folder 3_Checking_data into pandas dataframes called trip, weather and station. (Don't forget to import the pandas library first!) <br>
Create three variables called trip_duplicates_num, weather_duplicates_num and station_duplicatess_num that contains how many duplicated rows are in each dataframe.

In [1]:
import pandas as pd

## Import datasets

# trip
trip = pd.read_csv('trip.csv')
#count no of duplicate rows
trip_duplicates_num = trip.duplicated().sum()

# weather
weather = pd.read_csv('weather.csv')
#count no of duplicate rows
weather_duplicates_num = weather.duplicated().sum()

# station
station = pd.read_csv('station.csv')
#count no of duplicate rows
station_duplicates_num = station.duplicated().sum()

### 1 - check yourself

In [2]:
print('Length of trip dataframe should be 144115 with 100 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(trip), trip_duplicates_num))
print('Length of weather dataframe should be 928 with 8 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(weather), weather_duplicates_num))
print('Length of station dataframe should be 69 with 0 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(station), station_duplicates_num))

Length of trip dataframe should be 144115 with 100 duplicated rows.
The length of your dataframe is 144115 and you counted 100 duplicated rows

Length of weather dataframe should be 928 with 8 duplicated rows.
The length of your dataframe is 928 and you counted 8 duplicated rows

Length of station dataframe should be 69 with 0 duplicated rows.
The length of your dataframe is 69 and you counted 0 duplicated rows



### 2 - exercise
For all 3 dataframes delete the duplicated rows in place (without creating a new dataframe) <br>

In [3]:
# Remove duplicated rows in each dataframe

#trip
trip.drop_duplicates(inplace = True)

#weather
weather.drop_duplicates(inplace = True)

#station
station.drop_duplicates(inplace = True)

### 2 - check yourself

In [4]:
print('Length of trip dataframe should be 144015.\n\
The length of your dataframe is {}\n'.format(len(trip)))
print('Length of weather dataframe should be 920.\n\
The length of your dataframe is {}\n'.format(len(weather)))
print('Length of station dataframe should be 69.\n\
The length of your dataframe is {}\n'.format(len(station)))

Length of trip dataframe should be 144015.
The length of your dataframe is 144015

Length of weather dataframe should be 920.
The length of your dataframe is 920

Length of station dataframe should be 69.
The length of your dataframe is 69



### 3 - exercise
For all dataframes check if there are columns with missing values. <br>
Create 3 lists called trip_columns_with_missing_data, weather_columns_with_missing_data, station_columns_with_missing_data that contains the names of the columns with missing values of the dataframe. <br>
You can populate these lists by hand, or as an advanced task you can use pandas methods.

In [5]:
# Check if columns in trip df has any missing values
trip.isnull().sum()
# Save the column names which have missing values
trip_columns_with_missing_data = trip.columns[trip.isnull().any()].tolist()

# Check if columns in weather df has any missing values
weather.isnull().sum()
# Save the column names which have missing values
weather_columns_with_missing_data = weather.columns[weather.isnull().any()].tolist()

# Check if columns in station df has any missing values
station.isnull().sum()
# Save the column names which have missing values
station_columns_with_missing_data = station.columns[station.isnull().any()].tolist()

### 3 - check yourself

In [6]:
print('Columns with missing values in the trip dataframe are:\n \
Subscription Type\n\
You have found:\n {}\n'.format(','.join(trip_columns_with_missing_data)))

print('Columns with missing values in the weather dataframe are:\n \
Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events\n\
You have found:\n {}\n'.format(','.join(weather_columns_with_missing_data)))

print('Columns with missing values in the station dataframe are:\n \
\n\
You have found:\n {}\n'.format(','.join(station_columns_with_missing_data)))

Columns with missing values in the trip dataframe are:
 Subscription Type
You have found:
 Subscription Type

Columns with missing values in the weather dataframe are:
 Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events
You have found:
 Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events

Columns with missing values in the station dataframe are:
 
You have found:
 



### 4 - exercise
How many values are missing in each column with missing values in the dataframes? <br>
Display the answer in any format you would like. As an advance task, try to display only the names of the columns that have missing values in them.

In [7]:
# Convert missing column names and their count into a dataframe
weather_missing_values = weather.isnull().sum()[weather.isnull().sum() > 0]
weather_missing_values = pd.DataFrame(weather_missing_values)
weather_missing_values = weather_missing_values.rename_axis('column').reset_index()
weather_missing_values = weather_missing_values.rename(columns={0 : "count"})

trip_missing_values = trip.isnull().sum()[trip.isnull().sum() > 0]
trip_missing_values = pd.DataFrame(trip_missing_values)
trip_missing_values = trip_missing_values.rename_axis('column').reset_index()
trip_missing_values = trip_missing_values.rename(columns={0 : "count"})


In [8]:
for x,y in zip(weather_missing_values["column"],weather_missing_values["count"]):
    print('The number of missing values in ',x,'is',y)
    
for x,y in zip(trip_missing_values["column"],trip_missing_values["count"]):
    print('The number of missing values in ',x,'is',y)
    
# Station has no missing columns

The number of missing values in  Max_Temperature_F is 3
The number of missing values in  Mean_Temperature_F is 3
The number of missing values in  Min_TemperatureF is 3
The number of missing values in  Max_Gust_Speed_MPH is 138
The number of missing values in  Events is 782
The number of missing values in  Subscription Type is 10


In [9]:
for x,y in zip(weather_columns_with_missing_data,weather_missing_values):
    print('The number of missing values in ',x,'is',y)
    
for x,y in zip(trip_columns_with_missing_data,trip_missing_values):
    print('The number of missing values in ',x,'is',y)

The number of missing values in  Max_Temperature_F is column
The number of missing values in  Mean_Temperature_F is count
The number of missing values in  Subscription Type is column


### 4 - check yourself
The number of missing values in Max_Temperature_F column is 3 <br>
The number of missing values in Mean_Temperature_F column is 3<br>
The number of missing values in Min_TemperatureF column is 3<br>
The number of missing values in Max_Gust_Speed_MPH column is 138<br>
The number of missing values in Events column is 782<br>
The number of missing values in Subscription Type column is 10<br>

### 5 - exercise
Before deciding how to deal with the missing values, let's get more familiar with the data! <br>
 - How many values are in the columns with missing data? As an advanced task, try to display the number of rows for only these columns, not the others.
 - Display the mean of each numeric column in the weather dataframe. Which are 2 columns with the lowest mean? As an advanced task, try to display them in descending order!
 - And what about the Events column in the weather dataframe? What are the unique values and how many times do they occur?

In [10]:
# How many values are in the columns with missing data? As an advanced task, try to display the number of rows for only these columns, not the others.

print(trip.count()[trip.isnull().sum() > 0])
print(weather.count()[weather.isnull().sum() > 0])
# station has no columns with missing values

Subscription Type    144005
dtype: int64
Max_Temperature_F     917
Mean_Temperature_F    917
Min_TemperatureF      917
Max_Gust_Speed_MPH    782
Events                138
dtype: int64


In [11]:
# Display the mean of each numeric column in the weather dataframe. Which are 2 columns with the lowest mean?
# As an advanced task, try to display them in descending order!

weather.mean().sort_values(ascending = False).tail(2)

# Mean_Wind_Speed_MPH and Cloud_Cover have the lowest means

Mean_Wind_Speed_MPH     4.290217
Cloud_Cover             2.143478
dtype: float64

In [12]:
# What are the unique values in Events column in weather dataframe and how many times do they occur?
weather["Events"].value_counts()

Rain        101
Fog          34
rain          2
Fog-Rain      1
Name: Events, dtype: int64

### 5 - check yourself
The number of rows without missing data are:
- Max_Temperature_F column is 917
- Mean_Temperature_F column is 917
- Min_TemperatureF column is 917
- Max_Gust_Speed_MPH column is 782
- Events column is 138
- Subscription Type column is 144005

The columns with the lowest mean values are Cloud_clover and Mean_Wind_Speed_MPH <br><br>
In the Events column there are 101 rows with Rain, 34 rows with Fog, 2 rows with rain and 1 row with Fog-Rain

### 6 - exercise
So let's decide what we will do with the missing data!<br>
- In the Temperature columns, there are not too many missing data, so let's fill those cells with the mean of the column.
- In the Gust Speed column there are more than 10% of data missing so using the mean would be risky. We'll leave the missing values for now, and fill them later when using this dataframe.
- In the Events columns, the missing value means that there were no rain or fog that day. So let's fill those cells with the string 'no_event'
- In the Subscription type column, only a few rows are missing and we can't guess the original values. Let's delete those rows from the dataframe.

Create a new weather and trip dataframe called weather_filled and trip_filled where these solutions are applied!

In [13]:
# Take mean of columns which have missing values in weather dataframe
mean_Max_Temperature_F = weather['Max_Temperature_F'].mean()
mean_Min_TemperatureF = weather['Min_TemperatureF'].mean()
mean_Mean_Temperature_F = weather['Mean_Temperature_F'].mean()

# Fill missing values with mean in weather dataframe
weather_filled = weather.fillna(value={'Max_Temperature_F': mean_Max_Temperature_F,
                                       'Min_TemperatureF':mean_Min_TemperatureF,
                                       'Mean_Temperature_F':mean_Mean_Temperature_F,
                                        'Events': 'no_event'})

# Remoev rows with missing values from Subscription type column
trip_filled = trip.dropna(subset = ['Subscription Type'])

### 6 - check yourself

In [14]:
print('In the weather_filled dataframe the number of columns with missing data should be 1\n \
and in your dataframe the number is {}\n'.format(weather_filled.isnull().any().sum()))
print('The number of rows where the replacing value is not correct in the weather_filled dataframe: \n \
in column Max_Temperature_F is {} \n \
in column Mean_Temperature_F is {} \n \
in column Min_TemperatureF is {} \n \
in column Events is {} \n '.format((weather_filled[weather.Max_Temperature_F.isnull()]['Max_Temperature_F'] != weather.Max_Temperature_F.mean()).sum(),
                                             (weather_filled[weather.Mean_Temperature_F.isnull()]['Mean_Temperature_F'] != weather.Mean_Temperature_F.mean()).sum(),
                                             (weather_filled[weather.Min_TemperatureF.isnull()]['Min_TemperatureF'] != weather.Min_TemperatureF.mean()).sum(),
                                             (weather_filled[weather.Events.isnull()]['Events'] != 'no_event').sum()))
print('The length of the trip_filled dataframe should be 144005\n \
and in your dataframe number is {}\n'.format(len(trip_filled)))



In the weather_filled dataframe the number of columns with missing data should be 1
 and in your dataframe the number is 1

The number of rows where the replacing value is not correct in the weather_filled dataframe: 
 in column Max_Temperature_F is 0 
 in column Mean_Temperature_F is 0 
 in column Min_TemperatureF is 0 
 in column Events is 0 
 
The length of the trip_filled dataframe should be 144005
 and in your dataframe number is 144005



### 7 - exercise
Save these new dataframes into csv-s called weather_filled.csv and trip_filled.csv!

In [15]:
# Save weather_filled.csv to weather_filled.csv
weather_filled.to_csv('weather_filled.csv')

# Save trip_filled.csv to trip_filled.csv
trip_filled.to_csv('trip_filled.csv')


### 7 - check yourself

In [16]:
import os
if 'trip_filled.csv' in os.listdir():
    print('trip_filled.csv was successfully saved')
else:
    print('trip.csv was NOT successfully saved')
if 'weather_filled.csv' in os.listdir():
    print('weather_filled.csv was successfully saved')
else:
    print('weather_filled.csv was NOT successfully saved')

trip_filled.csv was successfully saved
weather_filled.csv was successfully saved


### +1 exercise

Let's repeat filling the missing values in the weather dataframe, but this time you'll first create the dictionary (beware, we haven't learnt it yet) used in the values part of the fillna() function and fill it automatically instead of typing the names of the columns: <br>
 - Create an empty dictionary
 - Loop through the columns of the weather dataframe
 - For each numeric column (int or float) that has missing values in it, but in less than 10% of the rows, create a key in the dictionary and assign the mean of the column as the value
 - For each string column that has missing values in it, the value in the dictionary should be the string 'no_event'
 - Create a dataframe called weather_filled2 and apply the fillna function on it, with the dictionary you created
 - Check if the weather filled2 dataframe is the same as the weather_filled dataframe

In [22]:
my_dict = {}

for col in weather:
    if weather[col].dtype == "int64" or "float64" and weather[col].dtype != "object":
        if weather[col].isna().sum() / weather[col].count() < 0.1:
            my_dict[col] = weather[col].mean()
    
    elif weather[col].dtype == "object":
        if weather[col].isnull().sum() > 0:
                my_dict[col] = "no_event"
            
weather_filled2 = weather.fillna(value = my_dict)
weather_filled2.sum() == weather_filled.sum()
            
    

Date                           True
Max_Temperature_F              True
Mean_Temperature_F             True
Min_TemperatureF               True
Max_Dew_Point_F                True
MeanDew_Point_F                True
Min_Dewpoint_F                 True
Max_Humidity                   True
Mean_Humidity                  True
Min_Humidity                   True
Max_Sea_Level_Pressure_In      True
Mean_Sea_Level_Pressure_In     True
Min_Sea_Level_Pressure_In      True
Max_Visibility_Miles           True
Mean_Visibility_Miles          True
Min_Visibility_Miles           True
Max_Wind_Speed_MPH             True
Mean_Wind_Speed_MPH            True
Max_Gust_Speed_MPH             True
Precipitation_In               True
Cloud_Cover                    True
Events                         True
Wind_Dir_Degrees               True
landmark                       True
dtype: bool